File: Arrays-and-Hashing/Medium/top-k-frequent-elements.java
//https://leetcode.com/problems/top-k-frequent-elements/description/
//N log n 

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        //hashmap of frequency
        Map<Integer,Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++){
            Integer value = map.get(nums[i]);

            if(value == null){
                map.put(nums[i], 1);
            }    

            else {
                map.put(nums[i], value + 1);
            }
        }

       

        //sort the entryset by value using a comparator
        TreeSet<Map.Entry<Integer,Integer>> sortedVals = new TreeSet<>(new ElementComparator());
        sortedVals.addAll(map.entrySet());

      
        //add top k values in sorted set to result
        int[] result = new int[k];
        int j = 0;
        
        for(Map.Entry<Integer,Integer> entry : sortedVals){
            if(j < k){
                result[j] = entry.getKey();
                j++;
            }
        }

        return result;

    }

    class ElementComparator implements Comparator<Map.Entry<Integer,Integer>>{
        public int compare(Map.Entry<Integer,Integer> e1, Map.Entry<Integer,Integer> e2) {
            //sort in descending order by values
            if(e2.getValue() != e1.getValue()){
            return e2.getValue() - e1.getValue();
            }
            else {return 
                e2.getKey() - e1.getKey();
            }
        }

    }


}

File: cpp/arrays-and-hashing/top-k-frequent-elements.cpp
#include <vector>
#include <unordered_map>

using namespace std;
class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        // get frequencies by hashmap

        std::unordered_map<int, int> hashMap;

        for (int i = 0; i < nums.size(); i++)
        {
            if (hashMap.find(nums[i]) == hashMap.end())
            {
                hashMap[nums[i]] = 1;
            }
            else
            {
                int value = hashMap[nums[i]];
                hashMap[nums[i]] = value + 1;
            }
        }

        // Create a vector of pairs from the unordered_map
        std::vector<std::pair<int, int>> vec(hashMap.begin(), hashMap.end());

        // Sort the vector based on the values (second element of the pair)
        std::sort(vec.begin(), vec.end(), [](const auto &a, const auto &b)
                  {
                      return a.second > b.second; // Sorts in descending order by values
                  });

        std::vector<int> result;
        // get top k elements
        for (int i = 0; i < k; i++)
        {
            result.push_back(vec[i].first);
        }

        return result;
    }
};